TestTxnCoordSenderCleanupOnAborted failing under stress

`make stress PKG=./pkg/kv TESTS=TestTxnCoordSenderCleanupOnAborted`

```
--- FAIL: TestTxnCoordSenderCleanupOnAborted (45.20s)
	txn_coord_sender_test.go:481: condition failed to evaluate within 45s: expected no heartbeat
		goroutine 41 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc42096fe00, 0x68b6980)
			/Users/andrei/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x68e84e0, 0xc4201c4780, 0xc42096fe00)
			/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:38 +0x7d
		github.com/cockroachdb/cockroach/pkg/kv.verifyCleanup(0xc4202527b0, 0x1, 0x8, 0x68eb6e0, 0xc420459980, 0xc4201c4780, 0xc42096fee8, 0x2, 0x2)
			/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:481 +0xda
		github.com/cockroachdb/cockroach/pkg/kv.TestTxnCoordSenderCleanupOnAborted(0xc4201c4780)
			/Users/andrei/work/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:678 +0x6d2
		testing.tRunner(0xc4201c4780, 0x5c39210)
			/Users/andrei/go/src/testing/testing.go:746 +0xd0
		created by testing.(*T).Run
			/Users/andrei/go/src/testing/testing.go:789 +0x2de
FAIL


ERROR: exit status 1

1946 runs completed, 1 failures, over 46s
context canceled
```

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -977,9 +977,11 @@ func (tc *TxnCoordSender) heartbeatLoop(ctx context.Context) {
 	var closer <-chan struct{}
 	{
 		tc.mu.Lock()
-		tc.mu.txnEnd = make(chan struct{})
 		closer = tc.mu.txnEnd
 		tc.mu.Unlock()
+		if closer == nil {
+			return
+		}
 	}
 	// Loop with ticker for periodic heartbeats.
 	for {
@@ -1298,6 +1300,9 @@ func (tc *TxnCoordSender) updateState(
 
 				// Only heartbeat the txn record if we're the root transaction.
 				if tc.typ == client.RootTxn {
+					// Create a channel to stop the heartbeat with the lock held
+					// to avoid a race between the async task and a subsequent commit.
+					tc.mu.txnEnd = make(chan struct{})
 					if err := tc.stopper.RunAsyncTask(
 						ctx, "kv.TxnCoordSender: heartbeat loop", func(ctx context.Context) {
 							tc.heartbeatLoop(ctx)

